### PR TITLE
fix: trim padding_out in conv_transpose x-backward (#4845)

### LINKED ISSUE #4845: Backward Pass Panics `.with_padding_out(!= 0)`
**Describe the bug**
Activation function (sigmoid or tanh) into Conv1dTranspose with `padding_out != 0` causes the gradients to mismatch the output in the backward pass, leading to a panic during `loss.backward()`. This happens across all backends. I have not tested with other activation functions or constellations.

**To Reproduce**
```rust
#[derive(Module, Debug)]
pub struct DecoderModel<B: Backend> {
    pub linear_1: Linear<B>,
    pub linear_2: Linear<B>,
    pub activation: Sigmoid,
    pub unflatten_shape: [usize; 2], // [channels, length]
    pub convt1: ConvTranspose1d<B>,
    pub convt2: ConvTranspose1d<B>,
}

impl<B: Backend> DecoderModel<B> {
    /// latent: [batch, latent_space_features]
    /// returns [batch, 2, 50] (or same as input)
    pub fn forward(&self, input: Tensor<B, 2>) -> Tensor<B, 3> {
        let x = self.linear_1.forward(input);
        let x = self.activation.forward(x);
        let x = self.linear_2.forward(x);
        let bsz = x.shape()[0];
        let unflat = x.reshape([bsz, self.unflatten_shape[0], self.unflatten_shape[1]]);
        let x = self.convt1.forward(unflat);
        let x = self.activation.forward(x);
        let x = self.convt2.forward(x);
        Tensor::cat(vec![x.clone(), x.slice_dim(2, 1)], 2) // shape: [batch, 2, 50]
    }
}

#[derive(Config, Debug)]
pub struct AutoencoderModelConfig {
    pub book_depth: usize, // Should be 25 for this use case.
    pub latent_space_features: usize,
    pub encoder_channels_1: usize,
    pub encoder_channels_2: usize,
    pub encoder_kernel_size: usize,
    pub decoder_kernel_size: usize,
    pub pool_kernel: usize,
    pub dropout: f64,
}

impl AutoencoderModelConfig {
    pub fn init<B: Backend>(&self, device: &B::Device) -> OrderBookSeriesAutoencoderModel<B> {
        let flatten_dim = self.encoder_channels_2 * ((self.book_depth * 2 - (self.encoder_kernel_size - 1) * 2) / self.pool_kernel);
        OrderBookSeriesAutoencoderModel {
            encoder: EncoderModel {
                conv1: Conv1dConfig::new(2, self.encoder_channels_1, self.encoder_kernel_size)
                    .with_padding(burn::nn::PaddingConfig1d::Explicit(0))
                    .init(device),
                conv2: Conv1dConfig::new(self.encoder_channels_1, self.encoder_channels_2, self.encoder_kernel_size)
                    .with_padding(burn::nn::PaddingConfig1d::Explicit(0))
                    .init(device),
                activation: Sigmoid::new(),
                dropout: DropoutConfig::new(self.dropout).init(),
                pool: MaxPool1dConfig::new(self.pool_kernel).with_stride(self.pool_kernel).init(),
                flatten_dim,
                linear_1: LinearConfig::new(flatten_dim, flatten_dim / 2).init(device),
                linear_2: LinearConfig::new(flatten_dim / 2, self.latent_space_features).init(device),
            },
            decoder: DecoderModel {
                linear_1: LinearConfig::new(self.latent_space_features, flatten_dim / 2).init(device),
                linear_2: LinearConfig::new(flatten_dim / 2, flatten_dim).init(device),
                activation: Sigmoid::new(),
                unflatten_shape: [self.encoder_channels_2, (self.book_depth * 2 - (self.encoder_kernel_size - 1) * 2) / self.pool_kernel],
                convt1: ConvTranspose1dConfig::new([self.encoder_channels_2, self.encoder_channels_1], self.decoder_kernel_size)
                    .with_padding(0)
                    .with_stride(self.pool_kernel)
                    .init(device),
                convt2: ConvTranspose1dConfig::new([self.encoder_channels_1, 2], self.decoder_kernel_size)
                    .with_padding(0)
                    //.with_stride(2)
                    .with_padding_out(0)
                    .init(device),
            },
        }
    }
}
```

**Expected behavior**
The backward pass should work properly and reliably, as it should granted the forward pass works.

**Desktop (please complete the following information):**
 - OS: Linux, openSUSE, Kernel 6.19.12-1-default
 - Browser: Headless Server
 - Version: 0.20.1

**Additional context**
Confirmed by having no problem (apart of curse of dimensionality) when using `.with_padding_out(0)`.

### PR DESCRIPTION
**Pull request by SAY-5**
### Checklist

- [x] Confirmed that `cargo run-checks` command has been executed.
- [ ] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

Fixes #4845.

### Changes

`conv_transpose{1,2,3}d_x_backward` ran the inverse conv on the full `output_grad`, but the trailing `padding_out` cells of a transpose-conv output do not depend on any input cell. The inverse conv therefore produced a result longer than the original input by `floor(padding_out / stride)` on each spatial axis, which then panicked when the gradient was registered against `x`.

Slice off the trailing cells after the inverse conv to recover the original input shape. Weight and bias gradients are unaffected (weight grad already slices to `weight_shape`; bias grad sums over spatial dims).

### Testing

- `cargo test -p burn-backend-tests --test autodiff conv` (98 conv autodiff tests pass, including the new regression test)
- `cargo test -p burn-backend-tests --test tensor conv_transpose` (19 tests pass)
- Added `test_conv_transpose1d_padding_out_stride1_backward_shape` mirroring the issue reproducer (stride=1, padding_out=1).

### CODE REVIEW

File: crates/burn-backend-tests/tests/autodiff/conv_transpose1d.rs [hunk 1]
```diff
@@ -211,6 +211,52 @@ fn test_conv_transpose1d_complex() {
     test.assert_grads(grads);
 }
 
+/// Regression test for #4845.
+///
+/// `ConvTranspose1d` with `padding_out != 0` and `stride == 1` used to panic in
+/// the backward pass because `conv_transpose1d_x_backward` did not account for
+/// the trailing `padding_out` cells, producing a gradient longer than `x`.
+#[test]
+fn test_conv_transpose1d_padding_out_stride1_backward_shape() {
```
Comment (laggui) [inline crates/burn-backend-tests/tests/autodiff/conv_transpose1d.rs:220]: The test should also validate the correctness, not just the shape. I think the current fix is incorrect (see my other comment).

You can use pytorch as a reference to compute the reference values.

File: crates/burn-backend/src/backend/ops/modules/conv.rs [hunk 2]
```diff
@@ -484,7 +484,37 @@ pub(crate) fn conv_transpose1d_x_backward<B: Backend>(
...
             options.groups,
         ),
-    )
+    );
+    // The forward transpose conv extends its output by `padding_out` zero cells
+    // on the trailing edge; those cells carry gradient but do not depend on
+    // any input cell. The inverse conv1d above therefore produces a result
+    // longer than the original `x` by `floor(padding_out / stride)`. Trim the
+    // trailing cells to recover the input length.
+    trim_padding_out_1d::<B>(grad, options.padding_out[0], options.stride[0])
```
Comment (laggui) [inline crates/burn-backend/src/backend/ops/modules/conv.rs]: /edit: nvm thanks for pointing that out!

**Review by laggui — CHANGES_REQUESTED** [review]
Thanks for addressing the issue!

**Review by laggui — APPROVED** [review]
